### ISSUE TITLE
Mismatch between train and eval loss results

### ISSUE THREAD

**Issue by Notou** [google-research/deeplab2#63]
Hello, 

I've trained a deeplabv3+ model on a custom dataset. It seems to be training properly, and the loss converges, in this case to about 0.03.
The issue is that when I run the evaluation, on the exact same data (not the evaluation set), the loss value returned is different from the previous one. 
More precisely, it goes from 'losses/train_total_loss': 0.032853287 to 'losses/eval_total_loss': 0.3922176
I figure that the train value is computed on just one loop (in this case 100 samples) and not the entire dataset, and there could be extreme cases that could skew the mean loss, but I would have seen huge loss spikes during training if that was the case.
So what I don't understand is how could the loss be so different between training and evaluation, on the same data?

As an additional note, I noticed that the more evaluation steps I make, the higher the returned loss is, but the IoU does not follow the inverse trend.

> restored model from /app/data/Modulation_recognition/models/Training_006/ckpt-70000.
> restored from checkpoint: /app/data/Modulation_recognition/models/Training_006/ckpt-70000
>  eval | step:  70000 | running 100 steps of evaluation...
>  eval | step:  70000 | eval time:   89.5 sec | output: 
>     {'evaluation/iou/IoU': 0.35476902,
>      'losses/eval_semantic_loss': 0.035643984,
>      'losses/eval_total_loss': 0.035643984}
> restoring or initializing model...
> restored model from /app/data/Modulation_recognition/models/Training_006/ckpt-70000.
> restored from checkpoint: /app/data/Modulation_recognition/models/Training_006/ckpt-70000
>  eval | step:  70000 | running 500 steps of evaluation...
>  eval | step:  70000 | eval time:  194.9 sec | output: 
>     {'evaluation/iou/IoU': 0.25706673,
>      'losses/eval_semantic_loss': 0.049148366,
>      'losses/eval_total_loss': 0.049148366}
> restoring or initializing model...
> restored model from /app/data/Modulation_recognition/models/Training_006/ckpt-70000.
> restored from checkpoint: /app/data/Modulation_recognition/models/Training_006/ckpt-70000
>  eval | step:  70000 | running 1000 steps of evaluation...
>  eval | step:  70000 | eval time:  249.3 sec | output: 
>     {'evaluation/iou/IoU': 0.5440319,
>      'losses/eval_semantic_loss': 0.13150834,
>      'losses/eval_total_loss': 0.13150834}
> restoring or initializing model...
> restored model from /app/data/Modulation_recognition/models/Training_006/ckpt-70000.
> restored from checkpoint: /app/data/Modulation_recognition/models/Training_006/ckpt-70000
>  eval | step:  70000 | running 10000 steps of evaluation...
>  eval | step:  70000 | eval time:  1709.6 sec | output: 
>     {'evaluation/iou/IoU': 0.73578227,
>      'losses/eval_semantic_loss': 0.23771212,
>      'losses/eval_total_loss': 0.23771212}

If it helps, here is the config file I'm using:
```
# Use a unique experiment_name for each experiment.
experiment_name: "Training_006"
model_options {
  # Update the path to the initial checkpoint (e.g., ImageNet
  # pretrained checkpoint).
  initial_checkpoint: ""
  backbone {
    name: "resnet50"
    output_stride: 32
  }
  decoder {
    feature_key: "res5"
    decoder_channels: 256
    aspp_channels: 256
    atrous_rates: 3
    atrous_rates: 6
    atrous_rates: 9
  }
  deeplab_v3_plus {
    low_level {
      feature_key: "res2"
      channels_project: 32
    }
    num_classes: 15
  }
}
trainer_options {
  save_checkpoints_steps: 1000
  save_summaries_steps: 100
  steps_per_loop: 100
  loss_options {
    semantic_loss {
      name: "softmax_cross_entropy"
      weight: 1.0
    }
  }
  solver_options {
    base_learning_rate: 0.0006
    training_number_of_steps: 30000
  }
}
train_dataset_options {
  dataset: "spawc_mod_reco"
  # Update the path to training set.
  file_pattern: "/app/data/Modulation_recognition/spawc21_wideband_dataset/tfrecords/train/*.tfrecord"
  # Adjust the batch_size accordingly to better fit your GPU/TPU memory.
  # Also see Q1 in g3doc/faq.md.
  batch_size: 4
  crop_size: 2049
  crop_size: 513
  # Skip resizing.
  min_resize_value: 0
  max_resize_value: 0
  augmentations {
    min_scale_factor: 1
    max_scale_factor: 1
    scale_factor_step_size: 1
  }
}
eval_dataset_options {
  dataset: "spawc_mod_reco"
  # Update the path to validation set.
  file_pattern: "/app/data/Modulation_recognition/spawc21_wideband_dataset/tfrecords/train/*.tfrecord"
  batch_size: 1
  crop_size: 2049
  crop_size: 600
  # Skip resizing.
  min_resize_value: 0
  max_resize_value: 0
} 
evaluator_options {
  eval_steps: 10000
  continuous_eval_timeout: -1
  save_predictions: true
  num_vis_samples: 200
  save_raw_predictions: false
}
```

Could it be something to do with the way the metrics are aggregated before printing? 

Thanks a lot

**Comment by Notou** [comment]
Hi @aquariusjay, thanks for getting back to me,

I asked the question because I'm getting this mismatch on the training dataset, not the evaluation one (that one also shows even higher loss, but that's a classic case of overfitting, I'm not surprisedby that bit)
So I don't see what could make what is essentially the same test function (the loss), operating on the exact same data, give different results depending on the way it is called.

**Comment by aquariusjay** [comment]
Closing the issue, as it is hard for us to figure out what could go wrong on the customized dataset.
Please feel free to reopen it if you find any issue related to the public datasets where our model zoo applies.
